### ISSUE TITLE
problem about vocabulary size

### ISSUE THREAD

**Issue by yuimo** [tensorflow/tensor2tensor#443]
hi, when i am running TranslateEnzhWmt8k, when i set the vocabulary size to 8K, the generated chinese vocabulary is normal, but when set it to 32K, the file contains lots of lines like this:
it seems like that some sentences are not split
is this normal?
can u give me some advice to resolve this?
thanks a lot.
<img width="370" alt="image" src="https://user-images.githubusercontent.com/22741826/33251027-71a7295a-d36f-11e7-8592-5ed1a034258e.png">

**Comment by martinpopel** [comment]
The idea of subwords/BPE is to merge frequent n-grams into one token, without limiting the n. In the original BPE paper, there is a restriction to not go over word boundaries - at least for languages with words separated by spaces and punctuation. However, there are different implementations, like https://github.com/google/sentencepiece, where space has no special meaning and no pre-tokenization (Moses tokenizer/MeCab/KyTea/Stanford) is needed, so there it is intentional that the resulting "sentence-piece" tokens may span several words. It is not obvious whether restricting the subword merges to not go over word boundaries is a good idea. You can try both for a given language and dataset and report the results (I don't remember I have seen such a comparison for Chinese).
I guess you know that for Chinese word tokenization, you need to apply some heuristics (e.g. the [Stanford implementation](https://nlp.stanford.edu/software/segmenter.shtml)) and this is not included in the current Tensor2tensor code base.

**Comment by twairball** [comment]
yes this is normal. `SubwordTokenizer.build_to_target_size` has `min_val` argument so you can set some threshold and this might remove some of these infrequent long subwords. 

In practice, I find that it doesn't matter (but smaller vocab can result in faster training)

**Comment by yuimo** [comment]
@twairball "In practice, I find that it doesn't matter "  you mean language like chinese without space between words?
i have run TranslateEnzhWmt8k as the original way implemented by the code, it's result are ok.
but as my "normal" way, i will segment the sentences to words seperated by space, then process it next. and then the result will be similary to english sentences
but as the implement of t2t, there is no need to segment first, you know the reason? or the result are similar?

**Comment by twairball** [comment]
Preprocessing with segmentation only affects the subword vocab, the transformer model isn't impacted by this. 

So if SubwordTextEncoder can make decent vocab without segmentation then the result is similar. 

At worst case, our Chinese vocab is just characters. With sufficient vocab size this is ok -- Chinese dictionary is only 23k characters with very sparse distribution. I think various experimenters on this repo will have found that Transformer's attention model seems very resilient to vocab encoding.

**Comment by rsepassi** [comment]
Thanks @martinpopel and @twairball!
